File: 1640_check_array_formation_through_concatenation/Solution.java
// LeetCode 1640 - Check Array Formation Through Concatenation
// https://leetcode.com/problems/check-array-formation-through-concatenation/

import java.util.*;

class Solution {
    public boolean canFormArray(int[] arr, int[][] pieces) {
        Map<Integer, int[]> byFirst = new HashMap<>();
        for (int[] p : pieces) byFirst.put(p[0], p);
        int i = 0;
        while (i < arr.length) {
            if (!byFirst.containsKey(arr[i])) return false;
            int[] p = byFirst.get(arr[i]);
            for (int x : p) {
                if (i >= arr.length || arr[i++] != x) return false;
            }
        }
        return true;
    }
}

File: 1640_check_array_formation_through_concatenation/solution.js
// LeetCode 1640 - Check Array Formation Through Concatenation
// https://leetcode.com/problems/check-array-formation-through-concatenation/

/**
 * @param {number[]} arr
 * @param {number[][]} pieces
 * @return {boolean}
 */
var canFormArray = function(arr, pieces) {
    const byFirst = new Map();
    for (const p of pieces) byFirst.set(p[0], p);
    let i = 0;
    while (i < arr.length) {
        if (!byFirst.has(arr[i])) return false;
        const p = byFirst.get(arr[i]);
        for (let j = 0; j < p.length; j++) if (arr[i + j] !== p[j]) return false;
        i += p.length;
    }
    return true;
};
